File: FateGrandAutomata.Core/Modules/AutoSkill.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using CoreAutomata;

namespace FateGrandAutomata
{
    public class AutoSkill
    {
        Dictionary<char, Action> _currentArray;
        readonly Dictionary<char, Action> _defaultFunctionArray,
            _startingMemberFunctionArray,
            _subMemberFunctionArray,
            _enemyTargetArray,
            _cardsPressed;

        public bool IsFinished { get; private set; }

        public bool NpsClicked { get; private set; }

        const int Normal = 1,
            Chaining = 2;

        public Battle Battle { get; private set; }
        public Card Card { get; private set; }

        void WaitForAnimationToFinish(TimeSpan? Timeout = null)
        {
            var img = ImageLocator.Battle;

            // slow devices need this. do not remove.
            Game.BattleScreenRegion.WaitVanish(img, TimeSpan.FromSeconds(2));

            Game.BattleScreenRegion.Exists(img, Timeout ?? TimeSpan.FromSeconds(5));
        }

        void CastSkill(Location Location)
        {
            Location.Click();

            if (Preferences.Instance.SkillConfirmation)
            {
                Game.BattleSkillOkClick.Click();
            }

            WaitForAnimationToFinish();
        }

        void SelectSkillTarget(Location Location)
        {
            Location.Click();

            AutomataApi.Wait(0.5);

            // Exit any extra menu
            Game.BattleExtrainfoWindowCloseClick.Click();

            WaitForAnimationToFinish();
        }

        void CastNoblePhantasm(Location Location)
        {
            if (!Battle.HasClickedAttack)
            {
                Battle.ClickAttack();

                // There is a delay after clicking attack before NP Cards come up. DON'T DELETE!
                AutomataApi.Wait(2);
            }

            /*
                Embed the PreloadNP Feature in this function to as a prerequisite for chaining to be viable.
               
                Problem with Chaining (as of right now). PreloadNP feature clicks cards prior to knowing what NPs are being clicked.
                
                If the NP was known (in this function for instance) then the script can register the face of the NP user first,
                then use that face to determine what cards to select in PreloadNP.
             */

            Location.Click();

            NpsClicked = true;
        }

        void OpenMasterSkillMenu()
        {
            Game.BattleMasterSkillOpenClick.Click();

            AutomataApi.Wait(0.5);
        }

        void CastMasterSkill(Location Location)
        {
            OpenMasterSkillMenu();

            CastSkill(Location);
        }

        void ChangeArray(Dictionary<char, Action> NewArray)
        {
            _currentArray = NewArray;
        }

        void BeginOrderChange()
        {
            OpenMasterSkillMenu();

            Game.BattleMasterSkill3Click.Click();

            if (Preferences.Instance.SkillConfirmation)
            {
                Game.BattleSkillOkClick.Click();
            }

            AutomataApi.Wait(0.3);

            ChangeArray(_startingMemberFunctionArray);
        }

        void SelectStartingMember(Location Location)
        {
            Location.Click();

            ChangeArray(_subMemberFunctionArray);
        }

        void SelectSubMemeber(Location Location)
        {
            Location.Click();

            AutomataApi.Wait(0.3);

            Game.BattleOrderChangeOkClick.Click();

            // Extra wait to allow order change dialog to close
            AutomataApi.Wait(1);

            WaitForAnimationToFinish(TimeSpan.FromSeconds(15));

            // Extra wait for the lag introduced by Order change
            AutomataApi.Wait(1);

            ChangeArray(_defaultFunctionArray);
        }

        void SelectTarget()
        {
            ChangeArray(_enemyTargetArray);
        }

        void SelectEnemyTarget(Location Location)
        {
            Location.Click();

            AutomataApi.Wait(0.5);

            // Exit any extra menu
            Game.BattleExtrainfoWindowCloseClick.Click();

            ChangeArray(_defaultFunctionArray);
        }

        void PreloadNp()
        {
            if (!Battle.HasClickedAttack)
            {
                Battle.ClickAttack();

                // There is a delay after clicking attack before NP Cards come up. DON'T DELETE!
                AutomataApi.Wait(2);
            }

            ChangeArray(_cardsPressed);
        }

        void PressCards(int NoOfCards)
        {
            Card.ClickCommandCards(NoOfCards);

            ChangeArray(_defaultFunctionArray);
        }

        public AutoSkill()
        {
            _defaultFunctionArray = new Dictionary<char, Action>
            {
                ['a'] = () => CastSkill(Game.BattleSkill1Click),
                ['b'] = () => CastSkill(Game.BattleSkill2Click),
                ['c'] = () => CastSkill(Game.BattleSkill3Click),
                ['d'] = () => CastSkill(Game.BattleSkill4Click),
                ['e'] = () => CastSkill(Game.BattleSkill5Click),
                ['f'] = () => CastSkill(Game.BattleSkill6Click),
                ['g'] = () => CastSkill(Game.BattleSkill7Click),
                ['h'] = () => CastSkill(Game.BattleSkill8Click),
                ['i'] = () => CastSkill(Game.BattleSkill9Click),

                ['j'] = () => CastMasterSkill(Game.BattleMasterSkill1Click),
                ['k'] = () => CastMasterSkill(Game.BattleMasterSkill2Click),
                ['l'] = () => CastMasterSkill(Game.BattleMasterSkill3Click),

                ['x'] = BeginOrderChange,
                ['t'] = SelectTarget,
                ['n'] = PreloadNp,

                ['0'] = () => { },

                ['1'] = () => SelectSkillTarget(Game.BattleServant1Click),
                ['2'] = () => SelectSkillTarget(Game.BattleServant2Click),
                ['3'] = () => SelectSkillTarget(Game.BattleServant3Click),

                ['4'] = () => CastNoblePhantasm(Game.BattleNpCardClickArray[0]),
                ['5'] = () => CastNoblePhantasm(Game.BattleNpCardClickArray[1]),
                ['6'] = () => CastNoblePhantasm(Game.BattleNpCardClickArray[2])
            };

            _startingMemberFunctionArray = new Dictionary<char, Action>
            {
                ['1'] = () => SelectStartingMember(Game.BattleStartingMember1Click),
                ['2'] = () => SelectStartingMember(Game.BattleStartingMember2Click),
                ['3'] = () => SelectStartingMember(Game.BattleStartingMember3Click)
            };

            _subMemberFunctionArray = new Dictionary<char, Action>
            {
                ['1'] = () => SelectSubMemeber(Game.BattleSubMember1Click),
                ['2'] = () => SelectSubMemeber(Game.BattleSubMember2Click),
                ['3'] = () => SelectSubMemeber(Game.BattleSubMember3Click)
            };

            _enemyTargetArray = new Dictionary<char, Action>
            {
                ['1'] = () => SelectEnemyTarget(Game.BattleTargetClickArray[0]),
                ['2'] = () => SelectEnemyTarget(Game.BattleTargetClickArray[1]),
                ['3'] = () => SelectEnemyTarget(Game.BattleTargetClickArray[2]),
            };

            _cardsPressed = new Dictionary<char, Action>
            {
                ['1'] = () => PressCards(1),
                ['2'] = () => PressCards(2)
            };
        }

        readonly List<List<string>> _commandTable = new List<List<string>>();

        void InitCommands()
        {
            var stageCount = 0;

            foreach (var commandList in Preferences.Instance.SkillCommand.Split(','))
            {
                if (Regex.IsMatch(commandList, @"[^0]"))
                {
                    if (Regex.IsMatch(commandList, @"^[1-3]"))
                    {
                        throw new ScriptExitException($"Error at '{commandList}': Skill Command cannot start with number '1', '2' and '3'!");
                    }

                    if (Regex.IsMatch(commandList, @"[^,]#") || Regex.IsMatch(commandList, @"#[^,]"))
                    {
                        throw new ScriptExitException($"Error at '{commandList}': '#' must be preceded and followed by ','! Correct: ',#,'");
                    }

                    if (Regex.IsMatch(commandList, @"[^a-l1-6#ntx]"))
                    {
                        throw new ScriptExitException($"Error at '{commandList}': Skill Command exceeded alphanumeric range! Expected 'x', 'n', 't' or range 'a' to 'l' for alphabets and '0' to '6' for numbers.");
                    }
                }

                if (stageCount >= _commandTable.Count)
                {
                    _commandTable.Add(new List<string>());
                }

                if (commandList == "#")
                {
                    ++stageCount;
                }
                else _commandTable[stageCount].Add(commandList);
            }
        }

        public void ResetState()
        {
            IsFinished = !Preferences.Instance.EnableAutoSkill;

            ChangeArray(_defaultFunctionArray);
        }

        public void Init(Battle BattleModule, Card CardModule)
        {
            Battle = BattleModule;
            Card = CardModule;

            if (Preferences.Instance.EnableAutoSkill)
            {
                InitCommands();
            }

            ResetState();
        }

        string GetCommandListFor(int Stage, int Turn)
        {
            if (Stage < _commandTable.Count)
            {
                var commandList = _commandTable[Stage];

                if (Turn < commandList.Count)
                {
                    return commandList[Turn];
                }
            }

            return null;
        }

        void ExecuteCommandList(string CommandList)
        {
            foreach (var command in CommandList)
            {
                _currentArray[command]();
            }
        }

        public void ResetNpTimer()
        {
            NpsClicked = false;
        }

        public bool Execute()
        {
            var commandList = GetCommandListFor(Battle.CurrentStage, Battle.CurrentTurn);

            if (commandList != null)
            {
                ExecuteCommandList(commandList);
            }
            else if (Battle.CurrentStage + 1 >= _commandTable.Count)
            {
                // this will allow NP spam after all commands have been executed
                IsFinished = true;
            }

            return NpsClicked;
        }
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/scripts/modules/AutoSkill.kt
package com.mathewsachin.fategrandautomata.scripts.modules

import com.mathewsachin.fategrandautomata.core.*
import com.mathewsachin.fategrandautomata.scripts.ImageLocator
import com.mathewsachin.fategrandautomata.scripts.prefs.Preferences
import kotlin.time.Duration
import kotlin.time.seconds

typealias AutoSkillMap = Map<Char, () -> Unit>

class AutoSkill {
    private val defaultFunctionArray: AutoSkillMap = mapOf(
        'a' to { castSkill(Game.BattleSkill1Click) },
        'b' to { castSkill(Game.BattleSkill2Click) },
        'c' to { castSkill(Game.BattleSkill3Click) },
        'd' to { castSkill(Game.BattleSkill4Click) },
        'e' to { castSkill(Game.BattleSkill5Click) },
        'f' to { castSkill(Game.BattleSkill6Click) },
        'g' to { castSkill(Game.BattleSkill7Click) },
        'h' to { castSkill(Game.BattleSkill8Click) },
        'i' to { castSkill(Game.BattleSkill9Click) },

        'j' to { castMasterSkill(Game.BattleMasterSkill1Click) },
        'k' to { castMasterSkill(Game.BattleMasterSkill2Click) },
        'l' to { castMasterSkill(Game.BattleMasterSkill3Click) },

        'x' to { beginOrderChange() },
        't' to { selectTarget() },
        'n' to { useCommandCardsBeforeNp() },

        '0' to { },

        '1' to { selectSkillTarget(Game.BattleServant1Click) },
        '2' to { selectSkillTarget(Game.BattleServant2Click) },
        '3' to { selectSkillTarget(Game.BattleServant3Click) },

        '4' to { castNoblePhantasm(Game.BattleNpCardClickArray[0]) },
        '5' to { castNoblePhantasm(Game.BattleNpCardClickArray[1]) },
        '6' to { castNoblePhantasm(Game.BattleNpCardClickArray[2]) }
    )

    private val startingMemberFunctionArray: AutoSkillMap = mapOf(
        '1' to { selectStartingMember(Game.BattleStartingMember1Click) },
        '2' to { selectStartingMember(Game.BattleStartingMember2Click) },
        '3' to { selectStartingMember(Game.BattleStartingMember3Click) }
    )

    private val subMemberFunctionArray: AutoSkillMap = mapOf(
        '1' to { selectSubMember(Game.BattleSubMember1Click) },
        '2' to { selectSubMember(Game.BattleSubMember2Click) },
        '3' to { selectSubMember(Game.BattleSubMember3Click) }
    )

    private val enemyTargetArray: AutoSkillMap = mapOf(
        '1' to { selectEnemyTarget(Game.BattleTargetClickArray[0]) },
        '2' to { selectEnemyTarget(Game.BattleTargetClickArray[1]) },
        '3' to { selectEnemyTarget(Game.BattleTargetClickArray[2]) }
    )

    private val cardsPressedArray: AutoSkillMap = mapOf(
        '1' to { pressCards(1) },
        '2' to { pressCards(2) }
    )

    private var currentArray = defaultFunctionArray

    private lateinit var battle: Battle
    private lateinit var card: Card

    var isFinished = false
        private set

    var npsClicked = false
        private set

    private fun waitForAnimationToFinish(Timeout: Duration = 5.seconds) {
        val img = ImageLocator.Battle

        // slow devices need this. do not remove.
        Game.BattleScreenRegion.waitVanish(img, 2.seconds)

        Game.BattleScreenRegion.exists(img, Timeout)
    }

    private fun castSkill(Location: Location) {
        Location.click()

        if (Preferences.SkillConfirmation) {
            Game.BattleSkillOkClick.click()
        }

        waitForAnimationToFinish()
    }

    private fun selectSkillTarget(Location: Location) {
        Location.click()

        0.5.seconds.wait()

        // Exit any extra menu
        Game.BattleExtrainfoWindowCloseClick.click()

        waitForAnimationToFinish()
    }

    private fun castNoblePhantasm(Location: Location) {
        if (!battle.hasClickedAttack) {
            battle.clickAttack()

            // There is a delay after clicking attack before NP Cards come up. DON'T DELETE!
            2.seconds.wait()
        }

        Location.click()

        npsClicked = true
    }

    private fun openMasterSkillMenu() {
        Game.BattleMasterSkillOpenClick.click()

        0.5.seconds.wait()
    }

    private fun castMasterSkill(Location: Location) {
        openMasterSkillMenu()

        castSkill(Location)
    }

    private fun changeArray(NewArray: AutoSkillMap) {
        currentArray = NewArray
    }

    private fun beginOrderChange() {
        openMasterSkillMenu()

        Game.BattleMasterSkill3Click.click()

        if (Preferences.SkillConfirmation) {
            Game.BattleSkillOkClick.click()
        }

        0.3.seconds.wait()

        changeArray(startingMemberFunctionArray)
    }

    private fun selectStartingMember(Location: Location) {
        Location.click()

        changeArray(subMemberFunctionArray)
    }

    private fun selectSubMember(Location: Location) {
        Location.click()

        0.3.seconds.wait()

        Game.BattleOrderChangeOkClick.click()

        // Extra wait to allow order change dialog to close
        1.seconds.wait()

        waitForAnimationToFinish(15.seconds)

        // Extra wait for the lag introduced by Order change
        1.seconds.wait()

        changeArray(defaultFunctionArray)
    }

    private fun selectTarget() = changeArray(enemyTargetArray)

    private fun selectEnemyTarget(Location: Location) {
        Location.click()

        0.5.seconds.wait()

        // Exit any extra menu
        Game.BattleExtrainfoWindowCloseClick.click()

        changeArray(defaultFunctionArray)
    }

    private fun useCommandCardsBeforeNp() {
        if (!battle.hasClickedAttack) {
            battle.clickAttack()

            // There is a delay after clicking attack before NP Cards come up. DON'T DELETE!
            2.seconds.wait()
        }

        changeArray(cardsPressedArray)
    }

    private fun pressCards(NoOfCards: Int) {
        card.clickCommandCards(NoOfCards)

        changeArray(defaultFunctionArray)
    }

    fun resetState() {
        isFinished = !Preferences.EnableAutoSkill

        changeArray(defaultFunctionArray)
    }

    private val commandTable = mutableListOf<MutableList<String>>()

    private fun initCommands() {
        var stageCount = 0

        for (commandList in Preferences.SkillCommand.splitToSequence(',')) {
            if (commandList != "0") {
                if (Regex("""^[1-3]""").containsMatchIn(commandList)) {
                    throw ScriptExitException("Error at '${commandList}': Skill Command cannot start with number '1', '2' and '3'!")
                }

                if (Regex("""([^,]#)|(#[^,])""").containsMatchIn(commandList)) {
                    throw ScriptExitException("Error at '${commandList}': '#' must be preceded and followed by ','! Correct: ',#,'")
                }

                if (Regex("""[^a-l1-6#ntx]""").containsMatchIn(commandList)) {
                    throw ScriptExitException("Error at '${commandList}': Skill Command exceeded alphanumeric range! Expected 'x', 'n', 't' or range 'a' to 'l' for alphabets and '0' to '6' for numbers.")
                }
            }

            if (stageCount >= commandTable.size) {
                commandTable.add(mutableListOf())
            }

            if (commandList == "#") {
                ++stageCount
            } else commandTable[stageCount].add(commandList)
        }
    }

    fun init(BattleModule: Battle, CardModule: Card) {
        battle = BattleModule
        card = CardModule

        if (Preferences.EnableAutoSkill) {
            initCommands()
        }

        resetState()
    }

    private fun getCommandListFor(Stage: Int, Turn: Int): String {
        if (Stage < commandTable.size) {
            val commandList = commandTable[Stage]

            if (Turn < commandList.size) {
                return commandList[Turn]
            }
        }

        return ""
    }

    private fun executeCommandList(CommandList: String) {
        for (command in CommandList) {
            currentArray[command]?.invoke()
        }
    }

    fun execute(): Boolean {
        val commandList = getCommandListFor(battle.currentStage, battle.currentTurn)

        if (commandList.isNotEmpty()) {
            executeCommandList(commandList)
        } else if (battle.currentStage + 1 >= commandTable.size) {
            // this will allow NP spam after all commands have been executed
            isFinished = true
        }

        return npsClicked
    }

    fun resetNpTimer() {
        npsClicked = false
    }
}